File: java/src/main/java/org/twyszomirski/aoc/solutions/Day_7.java
package org.twyszomirski.aoc.solutions;


import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.function.BiFunction;
import java.util.stream.Collectors;

import static java.io.IO.println;

public class Day_7 {

    public static void solve() throws IOException {
        var lines = Files.readAllLines(Paths.get("src/main/resources/input_day_7.txt"));
        println("======== Day 7 ===========");
        part_1(lines);
        part_2(lines);
    }

    private static void part_1(List<String> lines) {
        var data = readData(lines);
        var sum = (Long) data.stream().filter(it -> isResultCorrect(it, List.of(Operation.ADD, Operation.MULTIPLY))).mapToLong(it -> it.result).sum();
        println(sum);
    }

    private static void part_2(List<String> lines) {
        var data = readData(lines);
        var sum = (Long) data.stream().filter(it -> isResultCorrect(it, List.of(Operation.ADD, Operation.MULTIPLY, Operation.CONCAT))).mapToLong(it -> it.result).sum();
        println(sum);
    }

    private static List<OperationData> readData(List<String> lines) {
        return lines.stream().map(it -> it.split(":")).map(it -> new OperationData(Long.valueOf(it[0]),
                Arrays.stream(it[1].split(" ")).filter(op -> !op.isBlank()).map(Long::valueOf).toList())).toList();
    }

    private static List<? extends List<Operation>> providePermutations(int length, List<Operation> variants) {
        if (length == 1) {
            return variants.stream().map(Collections::singletonList).map(ArrayList::new).toList();
        }

        var prev = providePermutations(length - 1, variants);
        return prev.stream().flatMap(p -> {
            return variants.stream().map(variant ->
            {
                var copy = new ArrayList<>(p);
                copy.add(variant);
                return copy;
            });
        }).toList();
    }

    private static boolean isResultCorrect(OperationData operationData, List<Operation> variants) {
        var permutations = providePermutations(operationData.operands.size() - 1, variants);
        return permutations.stream().anyMatch(operationData::addsUp);
    }

    record OperationData(Long result, List<Long> operands) {

        public boolean addsUp(List<Operation> operations) {
            var result = operands.getFirst();

            for (int i = 0; i < operations.size(); i++) {
                result = operations.get(i).apply(result, operands.get(i + 1));
            }
            return Objects.equals(result, this.result);
        }
    }

    enum Operation {
        ADD(Long::sum),
        MULTIPLY((a, b) -> a * b),
        CONCAT((a, b) -> Long.valueOf(a.toString() + b.toString()));

        private BiFunction<Long, Long, Long> op;

        private Operation(BiFunction<Long, Long, Long> op) {
            this.op = op;
        }

        public Long apply(Long a, Long b) {
            return op.apply(a, b);
        }
    }

}

File: kotlin/src/main/kotlin/org/twyszomirski/aoc/solutions/Day_7.kt
package org.twyszomirski.aoc.solutions

import java.io.File


class Day_7 {

    fun solve() {
        val lines = File("src/main/resources/input_day_7.txt").readLines()
        println("======== Day 7 ===========")
        part1(lines)
        part2(lines)
    }

    fun part1(input: List<String>) {
        val data = readData(input)
        val res = data.filter { testOperations(it, setOf(Symbol.SUM, Symbol.MULTIPLY))}.sumOf { it.first }
        println(res)
    }

    fun part2(input: List<String>) {
        val data = readData(input)
        val res = data.filter { testOperations(it, setOf(Symbol.SUM, Symbol.MULTIPLY, Symbol.CONCAT))}.sumOf { it.first }
        println(res)
    }

    fun readData(input: List<String>) = input.map {
        it.split(":")
    }.map { Pair(it[0].toLong(), it[1].split(" ").filter { it.isNotBlank() }.map { it.toLong() }) }


    fun testOperations(data: Pair<Long, List<Long>>, symbols: Set<Symbol>): Boolean {
        val combinations = createCombinations(data.second.size - 1, symbols)
        return combinations.any { c ->
            val result = data.second.drop(1).foldIndexed(data.second.first()) { index, acc, element ->
                c[index].process(acc, element)
            }
            result == data.first
        }
    }

    fun createCombinations(length: Int, symbols: Set<Symbol>): List<List<Symbol>> {
        var res: List<List<Symbol>> = symbols.map { listOf(it) }

        repeat(length - 1) {
            res = res.flatMap { pref ->
                symbols.map { sym ->
                    pref + sym
                }
            }
        }
        return res
    }

    enum class Symbol(val operation: (Long, Long) -> Long) {
        SUM ({ a, b -> a + b }),
        MULTIPLY({ a, b -> a * b }),
        CONCAT({ a, b -> (a.toString() + b.toString()).toLong() });

        fun process(a : Long, b: Long) = operation(a, b)
    }

}